File: kSmallestElementInSortedMatrix.java
// Time complexity - O(nlog(max - min)) Space - O(1)
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        if(matrix == null || matrix.length == 0 || k == 0){
            return 0;
        }
        int n = matrix.length;
        int low = matrix[0][0], high = matrix[n - 1][n - 1];
        
        while(low <= high){
            int mid = low + (high - low) / 2;
            
            int count = noLessThanEqualTo(matrix, mid);
            
            if(count < k){
                low = mid + 1;
            }else{
                high = mid - 1;
            }
        }
        
        
        return low;
    
    }
    
    private int noLessThanEqualTo(int[][] matrix, int target){
        int n = matrix.length;
        int i = n - 1, j = 0;
        int result = 0;
        while(i >= 0 && j < n){
            if(matrix[i][j] > target) i--;
            else{
                result += i + 1; j++;
            }
        }
        
        return result;
    }
}

File: k_smallest_element_in_sorted_matrix.py
# Time Complexity: O(k log n)
# Space Complexity: O(n) n - where n is minimum between n and k
import heapq


class Solution:
    #     We will add all the first n or k elements in the heap (whichever is minimum)
    #    Then we will heapify given heap, for every k we will
    #    pop the element from the heap and check if current elements
    #    row and columns are in the range and if then we will push the next child of
    #    current element in the heap(here as all the row and column both are sorted
    #    and we are adding 1st column values in the heap in start then we can consider next child as next column)
    #    and till k is positive we will do operations on heap and after that we will
    #    just pop the top element in the heap
    def kthSmallest(self, matrix, k):
        heapm = []
        n = len(matrix)
        for i in range(min(n, k)):
            heapq.heappush(heapm, (matrix[i][0], i, 0))

        heapq.heapify(heapm)

        while k:
            curr, r, c = heapq.heappop(heapm)

            if c < n - 1:
                heapq.heappush(heapm, (matrix[r][c + 1], r, c + 1))

            k -= 1

        return curr


